docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,404 @@
+
+    // Classifer The classifier on which packets should match
+    // to apply the NetworkQoS Rule.
+    // This field is optional, and in case it is not set the rule is applied
+    // to all egress traffic regardless of the destination.
+    // +optional
+    Classifier Classifer `json:"classifier"`
+
+    // +optional
+    Bandwidth Bandwidth `json:"bandwidth"`
+}
+
+type Classifer struct {
+    // +optional
+    To []Destination `json:"to"`
+
+    // +optional
+    Ports Ports `json:"ports"`
+}
+
+// Bandwidth controls the maximum of rate traffic that can be sent
+// or received on the matching packets.
+type Bandwidth struct {
+    // Rate The value of rate limit in kbps. Traffic over the limit
+    // will be dropped.
+    // +kubebuilder:validation:Minimum:=1
+    // +kubebuilder:validation:Maximum:=4294967295
+    // +optional
+    Rate uint32 `json:"rate"`
+
+    // Burst The value of burst rate limit in kilobits.
+    // This also needs rate to be specified.
+    // +kubebuilder:validation:Minimum:=1
+    // +kubebuilder:validation:Maximum:=4294967295
+    // +optional
+    Burst uint32 `json:"burst"`
+}
+
+// Ports specifies destination protocol and port on which NetworkQoS
+// rule is applied
+type Ports struct {
+    // protocol (tcp, udp, sctp) that the traffic must match.
+    // +kubebuilder:validation:Pattern=^TCP|UDP|SCTP$
+    // +optional
+    Protocol string `json:"protocol"`
+
+    // port that the traffic must match
+    // +kubebuilder:validation:Minimum:=1
+    // +kubebuilder:validation:Maximum:=65535
+    // +optional
+    Port int32 `json:"port"`
+}
+
+// Destination specifies either pod or CIDR on which NetworkQoS rule is applied.
+type Destination struct {
+    // +optional
+    PodSelector PodSelector `json:"podSelector,omitempty"`
+
+    // CIDR specifies the destination's CIDR. Only traffic heading
+    // to this CIDR will be marked with the DSCP value and bandwidth
+    // get applied.
+    // +optional
+    // +kubebuilder:validation:Format="cidr"
+    CIDR *string `json:"cidr,omitempty"`
+}
+
+// PodSelector chooses destination pod for the NetworkQoS rule, when namespace is not
+// specified, then it searches for pod on the NetworkQoS namespace.
+type PodSelector struct {
+    // Namespace name of the namespace on which pod resides.
+    // +optional
+    Namespace string `json:"namespace,omitempty"`
+
+    // NamespaceSelector label selector to select namespace on which pod resides.
+    // an alternative to using namespace.
+    // +optional
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+    // matchLabels is a map of {key,value} pairs to select destination pod.
+    // +optional
+    MatchLabels map[string]string `json:"matchLabels,omitempty"`
+
+    // matchExpressions is a list of label selector requirements to select destination pod.
+    // +optional
+    MatchExpressions []metav1.LabelSelectorRequirement `json:"matchExpressions,omitempty"`
+}
```

Comment:
this PodSelector seems very complicated.. 

Why can't we `mimic` what is specified in `networkpolicy.spec.ingress.from`
```
FIELDS:
  ipBlock	<IPBlock>
    ipBlock defines policy on a particular IPBlock. If this field is set then
    neither of the other fields can be.

  namespaceSelector	<LabelSelector>
    namespaceSelector selects namespaces using cluster-scoped labels. This field
    follows standard label selector semantics; if present but empty, it selects
    all namespaces.
    
    If podSelector is also set, then the NetworkPolicyPeer as a whole selects
    the pods matching podSelector in the namespaces selected by
    namespaceSelector. Otherwise it selects all pods in the namespaces selected
    by namespaceSelector.

  podSelector	<LabelSelector>
    podSelector is a label selector which selects pods. This field follows
    standard label selector semantics; if present but empty, it selects all
    pods.
    
    If namespaceSelector is also set, then the NetworkPolicyPeer as a whole
    selects the pods matching podSelector in the Namespaces selected by
    NamespaceSelector. Otherwise it selects the pods matching podSelector in the
    policy's own namespace.

```

Reply:
sure, it even simplifies peer config. changed it to use similar struct here as well.